In [4]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import matplotlib.pyplot as plt

import matplotlib as mpl
mpl.rcParams['font.sans-serif'] = ['SimHei']
mpl.rcParams['font.serif'] = ['SimHei']
import seaborn as sns
%matplotlib inline
p = sns.color_palette()
sns.set_style("darkgrid",{"font.sans-serif":['simhei', 'Arial']})

In [5]:
import os
data_dir = "D:/GRADUATE/BAIRONG/bairong_project/data/"
def csv_f(file_name):
    return os.path.join(data_dir,file_name)

als = pd.read_csv(csv_f("als_1568089567.csv"))

E:\Programmer\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3063: DtypeWarning: Columns (0,1,6,7,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131,132,133,134,135,136,137,138,139,140,141,142,143,144,145,146,147,148,149,150,151,152,153,154,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,172,173,174,175,176,177,178,179,180,181,182,183,184,185,186,187,188,189,190,191,192,193,194,195,196,197,198,199,200,201,202,203,204,205,206,207,208,209,210,211,212,213,214,215,216,217,218,219,220,221,222,223,224,225,226,227,228,229,230,231,232,233,234,235,236,237,238,239,240,241,242,243,244,245,246,247,248,249,250,251,252,253,254,2

In [6]:
# 自定义转换函数

def standardize(value):
    """
    转换字符串/float等数字为整型
    """
    if not pd.isnull(value):
        #print(value)
        new_value = np.float(value)
        # 如果是float的字符串，会出现错误，
        return new_value
    else:
        return value

# als_.iloc[:,12].apply(standardize).value_counts(dropna=False)

In [7]:
als_ = als.iloc[1:,:]

In [8]:
als_df = pd.DataFrame(columns=als_.columns.values)

In [9]:
for i in range(0,13):
    als_df.iloc[:,i] = als_.iloc[:,i]

In [10]:
for i in range(13,len(als_.columns.values)):
    als_df.iloc[:,i] =  als_.iloc[:,i].apply(standardize)

In [11]:
als_df

cus_num other_var1                  id         cell name   user_date  \
1           4         01  13040219740924##2#  1833006####   李#  2017-12-08   
2           1         01  13032419870423##1#  1333333####  宁##  2017-04-26   
3           3         01  13040219640128##2#  1883106####  冉##  2017-10-30   
4           2         01  13032419940315##1#  1580117####  闫##  2017-12-04   
5           5         01  13040219800914##2#  1373004####   张#  2017-08-14   
...       ...        ...                 ...          ...  ...         ...   
93255     99b        NaN  11010919900307##2#  1381094####  刘##  2017-07-24   
93256     99a        NaN  12022519940420##1#  1761228####  孟##  2017-07-13   
93257    999a        NaN  13092119861024##1#  1509863####  贾##  2017-07-22   
93258      9a        NaN  11010519941013##4#  1352036####  王##  2017-07-06   
93259    999b        NaN  13072919960228##3#  1512733####  杜##  2017-07-06   

      flagy custApiCode sl_user_date                 swift_number  ...  \
1         1     3000983   2017-12-08  4000056_20190910122607_5863  ...   
2         1     3000983   2017-04-26  4000056_20190910122607_5864  ...   
3         1     3000983   2017-10-30  4000056_20190910122607_4895  ...   
4         1     3000983   2017-12-04  4000056_20190910122607_1814  ...   
5         1     3000983   2017-08-14  4000056_20190910122607_5865  ...   
...     ...         ...          ...                          ...  ...   
93255     0     3100122   2017-07-24  4000056_20190910124501_0086  ...   
93256     0     3100122   2017-07-13  4000056_20190910124501_3283  ...   
93257     0     3100122   2017-07-22  4000056_20190910124501_0861  ...   
93258     0     3100122   2017-07-06  4000056_20190910124501_3284  ...   
93259     0     3100122   2017-07-06  4000056_20190910124501_3282  ...   

      als_lst_id_bank_csinteday als_lst_id_nbank_inteday  \
1                           NaN                    112.0   
2                           NaN                     62.0   
3                           NaN                      2.0   
4                           NaN                      6.0   
5                           NaN                      5.0   
...                         ...                      ...   
93255                       NaN                      NaN   
93256                       NaN                      NaN   
93257                       1.0                    305.0   
93258                       NaN                      9.0   
93259                       1.0                      1.0   

      als_lst_id_nbank_consnum  als_lst_id_nbank_csinteday  \
1                          1.0                         1.0   
2                          1.0                         1.0   
3                          1.0                         1.0   
4                          1.0                         1.0   
5                          1.0                         1.0   
...                        ...                         ...   
93255                      NaN                         NaN   
93256                      NaN                         NaN   
93257                      1.0                         1.0   
93258                      1.0                         1.0   
93259                      1.0                         1.0   

       als_lst_cell_bank_inteday  als_lst_cell_bank_consnum  \
1                            NaN                        NaN   
2                            NaN                        NaN   
3                            NaN                        NaN   
4                            NaN                        NaN   
5                            NaN                        NaN   
...                          ...                        ...   
93255                        NaN                        NaN   
93256                        NaN                        NaN   
93257                        4.0                        1.0   
93258                        NaN                        NaN   
93259                

In [12]:
als_df.iloc[:,15].value_counts(dropna=False).sort_index()

1.0    10292
2.0     2936
3.0      868
4.0      212
5.0       47
6.0        5
7.0        4
NaN    78895
Name: als_d7_id_caon_allnum, dtype: int64

In [13]:
import json

als_col = als_df.columns.values
stat = dict()
for i in range(13,len(als_df.columns.values)):
    json_str = json.dumps((als_df.iloc[1:,i].value_counts(dropna=False).sort_index()).to_dict())
    stat[als_col[i]] = json_str

    
##保存文件
import csv

with open("als_temp.csv","w",newline='') as csv_file:
	writer = csv.writer(csv_file)
	for key,value in stat.items():
		writer.writerow([key,value])

In [14]:
# 统计缺失值

In [15]:
pd.DataFrame(als_df.isna().sum()).to_csv("als_nan.csv")

In [17]:
len(als_df)

93259

# 处理alm_0 3 6 9数据
解释：
alm-0   2017-10-30   表示最近的违约发生的时间（如有）
alm-3 	2017-07-30	 表示3个月前
alm-6 	2017-04-30
alm-9   2017-01-30

In [18]:
alm_0 = pd.read_csv(csv_f("alm_0.csv"))
alm_3 = pd.read_csv(csv_f("alm_3.csv"))
alm_6 = pd.read_csv(csv_f("alm_6.csv"))
alm_9 = pd.read_csv(csv_f("alm_9.csv"))

E:\Programmer\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3063: DtypeWarning: Columns (1,5,6,10,11,12,13,14,16,17,18,19,20,21,22,24,25,26,27,28,31,32,33,34,35,36,37,39,40,41,42,43,44,45,46,47,48,49,50,51,53,54,55,56,57,58) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [19]:
alm_3.iloc[0,:]

Unnamed: 0                                  0
cus_num                               cus_num
id                                         id
cell                                     cell
name                                     name
flagy                                   flagy
custApiCode                       custApiCode
user_date                           user_date
swift_number                              流水号
cus_username                             客户账号
code                                      返回码
flag_media_c                           产品输出标识
media_tot_m3_visitdays           近3个月所有类目浏览等级
media_tot_m3_catenum               近3个月浏览类目数量
media_max_m3_days               近3个月最大单类目浏览等级
media_max_m3_cate               近3个月最大浏览等级的类目
media_max_m3_days_prop         近3个月最大浏览等级类目占比
media_m3_CJ_visitdays           近3个月财经信息总浏览等级
media_m3_WXYS_visitdays       近3个月文学艺术信息总浏览等级
media_m3_MYYE_visitdays      近3个月母婴/育儿信息总浏览等级
media_tot_m12_visitdays         近12个月所有类目浏览等级
media_tot_m12_catenum             

In [20]:
alm_0

Unnamed: 0 cus_num                  id         cell name  flagy  \
0               0       1  13032419870423##1#  1333333####  宁##      1   
1               1      10  13040219940711##2#  1508100####  蔺##      1   
2               2     100  13043119850415##3#  1383208####   桑#      1   
3               3    1000  13092519900220##2#  1346376####  刘##      1   
4               4   10000  14270319660329##4#  1363359####  吴##      0   
...           ...     ...                 ...          ...  ...    ...   
88100       88100    7869  32083119931014##3#  1865240####  闵##      0   
88101       88101   78690  61032619910208##6#  1839276####  刘##      0   
88102       88102   78691  61010419990302##1#  1336399####  张##      0   
88103       88103   78692  45092319900504##2#  1859394####  刘##      1   
88104       88104   78693  42900520000719##1#  1354596####  肖##      0   

       custApiCode   user_date                 swift_number    cus_username  \
0          3000983  2017-04-26  4000056_20190606101948_6234  yejingyou_test   
1          3000983  2017-10-26  4000056_20190606101948_3022  yejingyou_test   
2          3000983  2017-12-09  4000056_20190606101952_4792  yejingyou_test   
3          3000983  2017-12-05  4000056_20190606102032_3642  yejingyou_test   
4          3000983  2017-05-10  4000056_20190606102626_0291  yejingyou_test   
...            ...         ...                          ...             ...   
88100      3000983  2018-01-24  4000056_20190606102504_9565  yejingyou_test   
88101      3000013  2018-08-09  4000056_20190606110921_3412  yejingyou_test   
88102      3000013  2018-08-09  4000056_20190606110921_3413  yejingyou_test   
88103      3000013  2018-08-09  4000056_20190606110921_3414  yejingyou_test   
88104      3000013  2018-08-09  4000056_20190606110921_0241  yejingyou_test   

       ...  cons_tot_m12_v_catenum  cons_max_m12_num  cons_max_m12_pay  \
0      ...                    24.0             183.0           30000.0   
1      ...                    19.0              35.0            3000.0   
2      ...                    17.0              10.0             600.0   
3      ...                    17.0             216.0           40000.0   
4      ...                     9.0              46.0            4000.0   
...    ...                     ...               ...               ...   
88100  ...                    17.0              78.0           20000.0   
88101  ...                    11.0               4.0             600.0   
88102  ...                     NaN               NaN               NaN   
88103  ...                     NaN               NaN               NaN   
88104  ...                    10.0               6.0            2000.0   

       cons_max_m12_paycate  cons_m12_RYBH_num cons_m12_RYBH_pay  \
0                       C11              183.0           10000.0   
1                        C7                2.0              20.0   
2                       C11                0.0               0.0   
3                        C7               11.0            2000.0   
4                       C14                NaN               NaN   
...                     ...                ...               ...   
88100                    C7                1.0              30.0   
88101                   C29                0.0               0.0   
88102                   NaN                NaN               NaN   
88103                   NaN                NaN               NaN   
88104                   C23                0.0               0.0   

       cons_m12_JYDQ_num  cons_m12_JYDQ_pay  cons_m12_MYYP_num  \
0                   52.0            30000.0               38.0   
1                    0.0                0.0               10.0   
2                    3.0              600.0                0.0   
3                    2.0              400.0                9.0   
4                    NaN                NaN               12.0   
...                  ...                ...                ...   
8

In [21]:
alm_0.iloc[:,5].value_counts()

0    73307
1    14798
Name: flagy, dtype: int64

In [22]:
alm_3_ = alm_3.iloc[1:,:]

In [23]:
alm_3_[alm_3_['cus_num']=='1']

Unnamed: 0 cus_num                  id         cell name flagy custApiCode  \
2           2       1  13032419870423##1#  1333333####  宁##     1     3000983   

    user_date                 swift_number    cus_username  ...  \
2  2017-01-26  4000056_20190612162224_0845  yejingyou_test  ...   

  cons_tot_m12_v_catenum cons_max_m12_num cons_max_m12_pay  \
2                     24              221            30000   

  cons_max_m12_paycate cons_m12_RYBH_num cons_m12_RYBH_pay cons_m12_JYDQ_num  \
2                  C11               221             20000                55   

  cons_m12_JYDQ_pay cons_m12_MYYP_num cons_m12_MYYP_pay  
2             30000                56              5000  

[1 rows x 59 columns]

In [24]:
alm_0[alm_0['cus_num']=='1']

Unnamed: 0 cus_num                  id         cell name  flagy  \
0           0       1  13032419870423##1#  1333333####  宁##      1   

   custApiCode   user_date                 swift_number    cus_username  ...  \
0      3000983  2017-04-26  4000056_20190606101948_6234  yejingyou_test  ...   

   cons_tot_m12_v_catenum  cons_max_m12_num  cons_max_m12_pay  \
0                    24.0             183.0           30000.0   

   cons_max_m12_paycate  cons_m12_RYBH_num cons_m12_RYBH_pay  \
0                   C11              183.0           10000.0   

   cons_m12_JYDQ_num  cons_m12_JYDQ_pay  cons_m12_MYYP_num  cons_m12_MYYP_pay  
0               52.0            30000.0               38.0             3000.0  

[1 rows x 59 columns]

In [25]:
alm_3_[alm_3_['cus_num']=='1']

Unnamed: 0 cus_num                  id         cell name flagy custApiCode  \
2           2       1  13032419870423##1#  1333333####  宁##     1     3000983   

    user_date                 swift_number    cus_username  ...  \
2  2017-01-26  4000056_20190612162224_0845  yejingyou_test  ...   

  cons_tot_m12_v_catenum cons_max_m12_num cons_max_m12_pay  \
2                     24              221            30000   

  cons_max_m12_paycate cons_m12_RYBH_num cons_m12_RYBH_pay cons_m12_JYDQ_num  \
2                  C11               221             20000                55   

  cons_m12_JYDQ_pay cons_m12_MYYP_num cons_m12_MYYP_pay  
2             30000                56              5000  

[1 rows x 59 columns]

In [26]:
# 按照
len(alm_0) - alm_0['cus_num'][alm_0['cus_num'].duplicated()].count()

88105

In [27]:
len(alm_3_) - alm_3_['cus_num'][alm_3_['cus_num'].duplicated()].count()

93259

In [28]:
len(alm_6) - alm_6['cus_num'][alm_6['cus_num'].duplicated()].count()

93260

In [29]:
len(alm_9) - alm_9['cus_num'][alm_9['cus_num'].duplicated()].count()

93260

In [30]:
# 按照cus_num来进行合并
# 查看是否对齐
alm_0_par = alm_0.loc[:,['cus_num','flagy','user_date']]
alm_0_par.rename(columns={'cus_num':'cus_num', 'flagy':'flagy_a','user_date':'user_date_a'}, inplace = True)

In [31]:
alm_0_par

cus_num  flagy_a user_date_a
0           1        1  2017-04-26
1          10        1  2017-10-26
2         100        1  2017-12-09
3        1000        1  2017-12-05
4       10000        0  2017-05-10
...       ...      ...         ...
88100    7869        0  2018-01-24
88101   78690        0  2018-08-09
88102   78691        0  2018-08-09
88103   78692        1  2018-08-09
88104   78693        0  2018-08-09

[88105 rows x 3 columns]

In [32]:
alm_3_par = alm_3.loc[:,['cus_num','flagy','user_date']]
alm_3_par.rename(columns={'cus_num':'cus_num', 'flagy':'flagy_b','user_date':'user_date_b'}, inplace = True)

In [33]:
alm_6_par = alm_6.loc[:,['cus_num','flagy','user_date']]
alm_6_par.rename(columns={'cus_num':'cus_num', 'flagy':'flagy_c','user_date':'user_date_c'}, inplace = True)

In [34]:
alm_6_par

cus_num flagy_c user_date_c
0      cus_num   flagy   user_date
1            3       1  2017-04-30
2            4       1  2017-06-08
3            2       1  2017-06-04
4            5       1  2017-02-14
...        ...     ...         ...
93255    9995b       0  2017-01-01
93256      99a       0  2017-01-13
93257       9a       0  2017-01-06
93258      99b       0  2017-01-24
93259    9996b       0  2017-01-04

[93260 rows x 3 columns]

In [35]:
alm_9_par = alm_9.loc[:,['cus_num','flagy','user_date']]
alm_9_par.rename(columns={'cus_num':'cus_num', 'flagy':'flagy_d','user_date':'user_date_d'}, inplace = True)

In [36]:
alm_9_par

cus_num flagy_d user_date_d
0      cus_num   flagy   user_date
1            3       1  2017-01-30
2            2       1  2017-03-04
3            4       1  2017-03-08
4            1       1  2016-07-26
...        ...     ...         ...
93255    9992a       0  2016-10-10
93256     999b       0  2016-10-06
93257      99a       0  2016-10-13
93258      99b       0  2016-10-24
93259       9a       0  2016-10-06

[93260 rows x 3 columns]

In [37]:
result = pd.merge(alm_0_par, alm_3_par, how='left', on=['cus_num'])

In [38]:
result = pd.merge(result, alm_6_par, how='left', on=['cus_num'])

In [39]:
result = pd.merge(result, alm_9_par, how='left', on=['cus_num'])

In [40]:
result

cus_num  flagy_a user_date_a flagy_b user_date_b flagy_c user_date_c  \
0           1        1  2017-04-26       1  2017-01-26       1  2016-10-26   
1          10        1  2017-10-26       1  2017-07-26       1  2017-04-26   
2         100        1  2017-12-09       1  2017-09-09       1  2017-06-09   
3        1000        1  2017-12-05       1  2017-09-05       1  2017-06-05   
4       10000        0  2017-05-10       0  2017-02-10       0  2016-11-10   
...       ...      ...         ...     ...         ...     ...         ...   
88100    7869        0  2018-01-24       0  2017-10-24       0  2017-07-24   
88101   78690        0  2018-08-09       0  2018-05-09       0  2018-02-09   
88102   78691        0  2018-08-09       0  2018-05-09       0  2018-02-09   
88103   78692        1  2018-08-09       1  2018-05-09       1  2018-02-09   
88104   78693        0  2018-08-09       0  2018-05-09       0  2018-02-09   

      flagy_d user_date_d  
0           1  2016-07-26  
1           1  2017-01-26  
2           1  2017-03-09  
3           1  2017-03-05  
4           0  2016-08-10  
...       ...         ...  
88100       0  2017-04-24  
88101       0  2017-11-09  
88102       0  2017-11-09  
88103       1  2017-11-09  
88104       0  2017-11-09  

[88105 rows x 9 columns]

In [41]:
# 从结果中判断flagy_i是否相等
result['cast'] = 0

In [42]:
result

cus_num  flagy_a user_date_a flagy_b user_date_b flagy_c user_date_c  \
0           1        1  2017-04-26       1  2017-01-26       1  2016-10-26   
1          10        1  2017-10-26       1  2017-07-26       1  2017-04-26   
2         100        1  2017-12-09       1  2017-09-09       1  2017-06-09   
3        1000        1  2017-12-05       1  2017-09-05       1  2017-06-05   
4       10000        0  2017-05-10       0  2017-02-10       0  2016-11-10   
...       ...      ...         ...     ...         ...     ...         ...   
88100    7869        0  2018-01-24       0  2017-10-24       0  2017-07-24   
88101   78690        0  2018-08-09       0  2018-05-09       0  2018-02-09   
88102   78691        0  2018-08-09       0  2018-05-09       0  2018-02-09   
88103   78692        1  2018-08-09       1  2018-05-09       1  2018-02-09   
88104   78693        0  2018-08-09       0  2018-05-09       0  2018-02-09   

      flagy_d user_date_d  cast  
0           1  2016-07-26     0  
1           1  2017-01-26     0  
2           1  2017-03-09     0  
3           1  2017-03-05     0  
4           0  2016-08-10     0  
...       ...         ...   ...  
88100       0  2017-04-24     0  
88101       0  2017-11-09     0  
88102       0  2017-11-09     0  
88103       1  2017-11-09     0  
88104       0  2017-11-09     0  

[88105 rows x 10 columns]

In [43]:
def cast(x):
    """根据多列的值做判断，生成新的列值"""
    if x['flagy_a'] == x['flagy_b'] and x['flagy_b']==x['flagy_c'] and x['flagy_c']==x['flagy_d']:
        return 1
    else:
        return -1
result['cast'] = result.apply(lambda x : cast(x),axis=1)

In [44]:
if result.loc[0,'flagy_a'] == result.loc[0,'flagy_b']:
    print("相等")
else:
    print("不等")

不等


In [45]:
result.loc[0,'flagy_b']

'1'

In [46]:
result.loc[0,'flagy_a']

1

In [47]:
# 统一数据类型
result['flagy_a'].dtype

dtype('int64')

In [48]:
result['flagy_b'].value_counts(dropna=False)

NaN    49152
0      21356
0      11855
1       3580
1       2162
Name: flagy_b, dtype: int64

In [49]:
result['flagy_c'].value_counts(dropna=False)

NaN    49152
0      21356
0      11855
1       3580
1       2162
Name: flagy_c, dtype: int64

In [50]:
result['flagy_d'].value_counts(dropna=False)

NaN    49152
0      21356
0      11855
1       3580
1       2162
Name: flagy_d, dtype: int64

In [51]:
result['flagy_b'] = result['flagy_b'].apply(lambda x : int(x) if not pd.isnull(x) else x)

In [52]:
result['flagy_c'] = result['flagy_c'].apply(lambda x : int(x) if not pd.isnull(x) else x)
result['flagy_d'] = result['flagy_d'].apply(lambda x : int(x) if not pd.isnull(x) else x)

In [53]:
result

cus_num  flagy_a user_date_a  flagy_b user_date_b  flagy_c user_date_c  \
0           1        1  2017-04-26      1.0  2017-01-26      1.0  2016-10-26   
1          10        1  2017-10-26      1.0  2017-07-26      1.0  2017-04-26   
2         100        1  2017-12-09      1.0  2017-09-09      1.0  2017-06-09   
3        1000        1  2017-12-05      1.0  2017-09-05      1.0  2017-06-05   
4       10000        0  2017-05-10      0.0  2017-02-10      0.0  2016-11-10   
...       ...      ...         ...      ...         ...      ...         ...   
88100    7869        0  2018-01-24      0.0  2017-10-24      0.0  2017-07-24   
88101   78690        0  2018-08-09      0.0  2018-05-09      0.0  2018-02-09   
88102   78691        0  2018-08-09      0.0  2018-05-09      0.0  2018-02-09   
88103   78692        1  2018-08-09      1.0  2018-05-09      1.0  2018-02-09   
88104   78693        0  2018-08-09      0.0  2018-05-09      0.0  2018-02-09   

       flagy_d user_date_d  cast  
0          1.0  2016-07-26    -1  
1          1.0  2017-01-26    -1  
2          1.0  2017-03-09    -1  
3          1.0  2017-03-05    -1  
4          0.0  2016-08-10    -1  
...        ...         ...   ...  
88100      0.0  2017-04-24    -1  
88101      0.0  2017-11-09     1  
88102      0.0  2017-11-09     1  
88103      1.0  2017-11-09     1  
88104      0.0  2017-11-09     1  

[88105 rows x 10 columns]

In [54]:
def cast(x):
    """根据多列的值做判断，生成新的列值"""
    if x['flagy_a'] == x['flagy_b'] and x['flagy_b']==x['flagy_c'] and x['flagy_c']==x['flagy_d']:
        return 1
    else:
        return 0
result['cast'] = result.apply(lambda x : cast(x),axis=1)

In [55]:
result

cus_num  flagy_a user_date_a  flagy_b user_date_b  flagy_c user_date_c  \
0           1        1  2017-04-26      1.0  2017-01-26      1.0  2016-10-26   
1          10        1  2017-10-26      1.0  2017-07-26      1.0  2017-04-26   
2         100        1  2017-12-09      1.0  2017-09-09      1.0  2017-06-09   
3        1000        1  2017-12-05      1.0  2017-09-05      1.0  2017-06-05   
4       10000        0  2017-05-10      0.0  2017-02-10      0.0  2016-11-10   
...       ...      ...         ...      ...         ...      ...         ...   
88100    7869        0  2018-01-24      0.0  2017-10-24      0.0  2017-07-24   
88101   78690        0  2018-08-09      0.0  2018-05-09      0.0  2018-02-09   
88102   78691        0  2018-08-09      0.0  2018-05-09      0.0  2018-02-09   
88103   78692        1  2018-08-09      1.0  2018-05-09      1.0  2018-02-09   
88104   78693        0  2018-08-09      0.0  2018-05-09      0.0  2018-02-09   

       flagy_d user_date_d  cast  
0          1.0  2016-07-26     1  
1          1.0  2017-01-26     1  
2          1.0  2017-03-09     1  
3          1.0  2017-03-05     1  
4          0.0  2016-08-10     1  
...        ...         ...   ...  
88100      0.0  2017-04-24     1  
88101      0.0  2017-11-09     1  
88102      0.0  2017-11-09     1  
88103      1.0  2017-11-09     1  
88104      0.0  2017-11-09     1  

[88105 rows x 10 columns]

In [56]:
result['cast'].value_counts()

0    49153
1    38952
Name: cast, dtype: int64

In [57]:
# 验证这些cast值为0的列就是那些flagy_b,flagy_c,flagy_d为Nan的列
result_nan_stat = result[result['cast']==0]
result_nan_stat

cus_num  flagy_a user_date_a  flagy_b user_date_b  flagy_c user_date_c  \
11941   16422        0  2018-05-08      NaN         NaN      NaN         NaN   
11942   16423        0  2018-03-15      NaN         NaN      NaN         NaN   
11944   16424        0  2018-05-12      NaN         NaN      NaN         NaN   
11945   16425        0  2018-07-16      NaN         NaN      NaN         NaN   
11946   16426        0  2018-03-10      NaN         NaN      NaN         NaN   
...       ...      ...         ...      ...         ...      ...         ...   
72651   65589        0  2018-03-20      NaN         NaN      NaN         NaN   
72654   65590        0  2018-03-20      NaN         NaN      NaN         NaN   
72655   65591        0  2018-03-20      NaN         NaN      NaN         NaN   
72656   65592        1  2018-03-20      1.0  2017-12-20      NaN         NaN   
72658   65594        1  2018-03-20      NaN         NaN      1.0  2017-09-20   

       flagy_d user_date_d  cast  
11941      NaN         NaN     0  
11942      NaN         NaN     0  
11944      NaN         NaN     0  
11945      NaN         NaN     0  
11946      NaN         NaN     0  
...        ...         ...   ...  
72651      NaN         NaN     0  
72654      NaN         NaN     0  
72655      NaN         NaN     0  
72656      NaN         NaN     0  
72658      1.0  2017-06-20     0  

[49153 rows x 10 columns]

In [58]:
result_nan_stat['flagy_b'].value_counts(dropna=False)

NaN    49152
1.0        1
Name: flagy_b, dtype: int64

In [59]:
result_nan_stat['flagy_c'].value_counts(dropna=False)

NaN    49152
1.0        1
Name: flagy_c, dtype: int64

In [60]:
result_nan_stat['flagy_d'].value_counts(dropna=False)

NaN    49152
1.0        1
Name: flagy_d, dtype: int64

### 经验证发现这些cast值为0的列就是那些flagy_b,flagy_c,flagy_d为Nan的列
### 而其他的38952都是完全相等的，因此可以认为alm_0的表的flagy状态就是基准，代表了其他3张表的flagy

In [61]:
alm_0

Unnamed: 0 cus_num                  id         cell name  flagy  \
0               0       1  13032419870423##1#  1333333####  宁##      1   
1               1      10  13040219940711##2#  1508100####  蔺##      1   
2               2     100  13043119850415##3#  1383208####   桑#      1   
3               3    1000  13092519900220##2#  1346376####  刘##      1   
4               4   10000  14270319660329##4#  1363359####  吴##      0   
...           ...     ...                 ...          ...  ...    ...   
88100       88100    7869  32083119931014##3#  1865240####  闵##      0   
88101       88101   78690  61032619910208##6#  1839276####  刘##      0   
88102       88102   78691  61010419990302##1#  1336399####  张##      0   
88103       88103   78692  45092319900504##2#  1859394####  刘##      1   
88104       88104   78693  42900520000719##1#  1354596####  肖##      0   

       custApiCode   user_date                 swift_number    cus_username  \
0          3000983  2017-04-26  4000056_20190606101948_6234  yejingyou_test   
1          3000983  2017-10-26  4000056_20190606101948_3022  yejingyou_test   
2          3000983  2017-12-09  4000056_20190606101952_4792  yejingyou_test   
3          3000983  2017-12-05  4000056_20190606102032_3642  yejingyou_test   
4          3000983  2017-05-10  4000056_20190606102626_0291  yejingyou_test   
...            ...         ...                          ...             ...   
88100      3000983  2018-01-24  4000056_20190606102504_9565  yejingyou_test   
88101      3000013  2018-08-09  4000056_20190606110921_3412  yejingyou_test   
88102      3000013  2018-08-09  4000056_20190606110921_3413  yejingyou_test   
88103      3000013  2018-08-09  4000056_20190606110921_3414  yejingyou_test   
88104      3000013  2018-08-09  4000056_20190606110921_0241  yejingyou_test   

       ...  cons_tot_m12_v_catenum  cons_max_m12_num  cons_max_m12_pay  \
0      ...                    24.0             183.0           30000.0   
1      ...                    19.0              35.0            3000.0   
2      ...                    17.0              10.0             600.0   
3      ...                    17.0             216.0           40000.0   
4      ...                     9.0              46.0            4000.0   
...    ...                     ...               ...               ...   
88100  ...                    17.0              78.0           20000.0   
88101  ...                    11.0               4.0             600.0   
88102  ...                     NaN               NaN               NaN   
88103  ...                     NaN               NaN               NaN   
88104  ...                    10.0               6.0            2000.0   

       cons_max_m12_paycate  cons_m12_RYBH_num cons_m12_RYBH_pay  \
0                       C11              183.0           10000.0   
1                        C7                2.0              20.0   
2                       C11                0.0               0.0   
3                        C7               11.0            2000.0   
4                       C14                NaN               NaN   
...                     ...                ...               ...   
88100                    C7                1.0              30.0   
88101                   C29                0.0               0.0   
88102                   NaN                NaN               NaN   
88103                   NaN                NaN               NaN   
88104                   C23                0.0               0.0   

       cons_m12_JYDQ_num  cons_m12_JYDQ_pay  cons_m12_MYYP_num  \
0                   52.0            30000.0               38.0   
1                    0.0                0.0               10.0   
2                    3.0              600.0                0.0   
3                    2.0              400.0                9.0   
4                    NaN                NaN               12.0   
...                  ...                ...                ...   
8

# 目标
### offset列名，将alm_3,alm_6,alm_9的数据形成时间序列
### 以alm_0中的第11~57列字段为基准，开始进行合并操作，形成列表


In [62]:
# 首先进行alm表的EDA，看看非profile字段(11~57列)数据的分布及Nan值情况
alm_0.iloc[:,12].value_counts(dropna=False)

NaN       85078
1.0         409
2.0         288
4.0         225
3.0         223
          ...  
545.0         1
93.0          1
187.0         1
1081.0        1
118.0         1
Name: media_tot_m3_visitdays, Length: 116, dtype: int64

In [67]:
# 统计一下NaN值和0值的多少
# 对alm_0的数据进行类型统一
alm_0.iloc[:,14].dtype

dtype('float64')

In [69]:
import json

alm_0_col = alm_0.columns.values
stat = dict()
for i in range(11,len(alm_0.columns.values)):
    json_str = json.dumps((alm_0.iloc[:,i].value_counts(dropna=False).sort_index()).to_dict())
    stat[alm_0_col[i]] = json_str

    
##保存文件
import csv

with open("alm_0_temp.csv","w",newline='') as csv_file:
	writer = csv.writer(csv_file)
	for key,value in stat.items():
		writer.writerow([key,value])

In [70]:
alm_0.isna().sum()

Unnamed: 0                      0
cus_num                         0
id                              0
cell                            0
name                           93
flagy                           0
custApiCode                     0
user_date                       0
swift_number                    0
cus_username                    0
code                            0
flag_media_c                    0
media_tot_m3_visitdays      85078
media_tot_m3_catenum        85078
media_max_m3_days           85078
media_max_m3_cate           85078
media_max_m3_days_prop      85078
media_m3_CJ_visitdays       86622
media_m3_WXYS_visitdays     87019
media_m3_MYYE_visitdays     87149
media_tot_m12_visitdays     85078
media_tot_m12_catenum       85078
media_max_m12_days          85078
media_max_m12_cate          85078
media_max_m12_days_prop     85078
media_m12_CJ_visitdays      86410
media_m12_WXYS_visitdays    86811
media_m12_MYYE_visitdays    86967
flag_consumption_c              0
cons_cont     

In [101]:
alm_0.iloc[:,13].value_counts(dropna=False).get('0')

In [106]:
alm_0_hollow_vc = pd.DataFrame(columns=alm_0_col)

In [189]:
zero_counts = {}
nan_counts = {}

for i in range(len(alm_0_col)):
    vc = alm_0.iloc[:,i].value_counts(dropna=False).sort_index().to_dict()
    zero_counts[alm_0_col[i]] = vc.get(0) if vc.get(0) else 0 + vc.get("0") if vc.get("0") else 0
    nan_counts[alm_0_col[i]] = alm_0.iloc[:,i].isna().sum()

In [190]:
zero_counts

{'Unnamed: 0': 1,
 'cus_num': 0,
 'id': 0,
 'cell': 0,
 'name': 0,
 'flagy': 73307,
 'custApiCode': 0,
 'user_date': 0,
 'swift_number': 0,
 'cus_username': 0,
 'code': 88105,
 'flag_media_c': 85078,
 'media_tot_m3_visitdays': 0,
 'media_tot_m3_catenum': 0,
 'media_max_m3_days': 0,
 'media_max_m3_cate': 0,
 'media_max_m3_days_prop': 0,
 'media_m3_CJ_visitdays': 0,
 'media_m3_WXYS_visitdays': 0,
 'media_m3_MYYE_visitdays': 0,
 'media_tot_m12_visitdays': 0,
 'media_tot_m12_catenum': 0,
 'media_max_m12_days': 0,
 'media_max_m12_cate': 0,
 'media_max_m12_days_prop': 0,
 'media_m12_CJ_visitdays': 0,
 'media_m12_WXYS_visitdays': 0,
 'media_m12_MYYE_visitdays': 0,
 'flag_consumption_c': 21301,
 'cons_cont': 0,
 'cons_time_recent': 0,
 'cons_tot_m3_visits': 22,
 'cons_tot_m3_num': 7641,
 'cons_tot_m3_pay': 8002,
 'cons_tot_m3_p_catenum': 7990,
 'cons_tot_m3_v_catenum': 22,
 'cons_max_m3_num': 7641,
 'cons_max_m3_pay': 12,
 'cons_max_m3_paycate': 0,
 'cons_m3_RYBH_num': 31275,
 'cons_m3_RYBH_pa

In [191]:
nan_counts

{'Unnamed: 0': 0,
 'cus_num': 0,
 'id': 0,
 'cell': 0,
 'name': 93,
 'flagy': 0,
 'custApiCode': 0,
 'user_date': 0,
 'swift_number': 0,
 'cus_username': 0,
 'code': 0,
 'flag_media_c': 0,
 'media_tot_m3_visitdays': 85078,
 'media_tot_m3_catenum': 85078,
 'media_max_m3_days': 85078,
 'media_max_m3_cate': 85078,
 'media_max_m3_days_prop': 85078,
 'media_m3_CJ_visitdays': 86622,
 'media_m3_WXYS_visitdays': 87019,
 'media_m3_MYYE_visitdays': 87149,
 'media_tot_m12_visitdays': 85078,
 'media_tot_m12_catenum': 85078,
 'media_max_m12_days': 85078,
 'media_max_m12_cate': 85078,
 'media_max_m12_days_prop': 85078,
 'media_m12_CJ_visitdays': 86410,
 'media_m12_WXYS_visitdays': 86811,
 'media_m12_MYYE_visitdays': 86967,
 'flag_consumption_c': 0,
 'cons_cont': 27563,
 'cons_time_recent': 21301,
 'cons_tot_m3_visits': 21301,
 'cons_tot_m3_num': 21301,
 'cons_tot_m3_pay': 21301,
 'cons_tot_m3_p_catenum': 21301,
 'cons_tot_m3_v_catenum': 21301,
 'cons_max_m3_num': 21301,
 'cons_max_m3_pay': 21301,
 '

In [155]:
temp = alm_0.loc[:,'media_tot_m3_catenum'].value_counts(dropna=False).sort_index().to_dict()

In [175]:
type(list(temp.keys())[-1])

float

In [183]:
temp2 = alm_0.loc[:,'media_max_m3_cate'].value_counts(dropna=False).sort_index().to_dict()

In [184]:
temp2.get(np.nan)

85078

In [188]:
alm_0.loc[:,'media_max_m3_cate'].isnull().sum()

85078

In [196]:
alm_0_hollow_vc = alm_0_hollow_vc.append(zero_counts,ignore_index=True)

In [203]:
alm_0_hollow_vc = alm_0_hollow_vc.append(nan_counts,ignore_index=True)

In [204]:
alm_0_hollow_vc

Unnamed: 0 cus_num id cell name  flagy custApiCode user_date swift_number  \
0          1       0  0    0    0  73307           0         0            0   
1          0       0  0    0   93      0           0         0            0   

  cus_username  ... cons_tot_m12_v_catenum cons_max_m12_num cons_max_m12_pay  \
0            0  ...                     16             1379               14   
1            0  ...                  21301            21301            21301   

  cons_max_m12_paycate cons_m12_RYBH_num cons_m12_RYBH_pay cons_m12_JYDQ_num  \
0                    0             27344             27365             30738   
1                21301             31373             31373             35346   

  cons_m12_JYDQ_pay cons_m12_MYYP_num cons_m12_MYYP_pay  
0             30738             26484             26484  
1             35346             33204             33204  

[2 rows x 59 columns]

In [212]:
alm_0_hollow_vc_T = alm_0_hollow_vc.T

In [213]:
alm_0_hollow_vc_T.columns=['zero_counts','nan_counts']

In [215]:
alm_0_hollow_vc_T.to_csv("alm_0_hollow_vc.csv")

In [217]:
alm_0.loc[:,'cons_m12_RYBH_num'].value_counts(dropna=False).sort_index()

0.0       27344
1.0        2379
2.0        4282
3.0        1779
4.0        3510
          ...  
1198.0        1
1388.0        1
1406.0        1
2775.0        1
NaN       31373
Name: cons_m12_RYBH_num, Length: 231, dtype: int64

In [218]:
len(alm_0)

88105

# 统计其他alm_3, alm_6, alm_9的值分布

In [235]:
# alm_3['cus_num' in alm_3['cus_num']]
alm_3[alm_3['cus_num'].isin(alm_0['cus_num'])]

Unnamed: 0 cus_num                  id         cell name flagy  \
1               1       3  13040219640128##2#  1883106####  冉##     1   
2               2       1  13032419870423##1#  1333333####  宁##     1   
3               3       2  13032419940315##1#  1580117####  闫##     1   
4               4       5  13040219800914##2#  1373004####   张#     1   
5               5       6  13040219820704##4#  1358158####   胡#     1   
...           ...     ...                 ...          ...  ...   ...   
91774       91774   7854b  35062419910225##1#  1885928####  许##     1   
91775       91775   7864b  35062719910818##2#  1330605####   杨#     0   
91776       91776   7868b  35062819810220##6#  1395969####  陈##     0   
91777       91777   7866b  35062819760723##1#  1516053####  庄##     0   
91780       91780   7867b  35062819800916##1#  1396262####  陈##     1   

      custApiCode   user_date                 swift_number    cus_username  \
1         3000983  2017-07-30  4000056_20190612162224_9236  yejingyou_test   
2         3000983  2017-01-26  4000056_20190612162224_0845  yejingyou_test   
3         3000983  2017-09-04  4000056_20190612162224_2185  yejingyou_test   
4         3000983  2017-05-14  4000056_20190612162224_3083  yejingyou_test   
5         3000983  2017-08-26  4000056_20190612162224_2187  yejingyou_test   
...           ...         ...                          ...             ...   
91774     3100122  2017-04-11  4000056_20190612170225_4877  yejingyou_test   
91775     3100122  2017-04-08  4000056_20190612170225_8347  yejingyou_test   
91776     3100122  2017-04-18  4000056_20190612170225_8349  yejingyou_test   
91777     3100122  2017-04-23  4000056_20190612170225_6153  yejingyou_test   
91780     3100122  2017-04-06  4000056_20190612170225_8348  yejingyou_test   

       ... cons_tot_m12_v_catenum cons_max_m12_num cons_max_m12_pay  \
1      ...                    NaN              NaN              NaN   
2      ...                     24              221            30000   
3      ...                      9               38            10000   
4      ...                     20              250            40000   
5      ...                     20               34             3000   
...    ...                    ...              ...              ...   
91774  ...                    NaN              NaN              NaN   
91775  ...                     20              148            30000   
91776  ...                     19               65            10000   
91777  ...                     13                8             3000   
91780  ...                     10               17             3000   

      cons_max_m12_paycate cons_m12_RYBH_num cons_m12_RYBH_pay  \
1                      NaN               NaN               NaN   
2                      C11               221             20000   
3                      C24               NaN               NaN   
4                       C7                 6                70   
5                       C7                 0                 0   
...                    ...               ...               ...   
91774                  NaN               NaN               NaN   
91775                   C7                12               900   
91776                   C7                28               500   
91777                  C26               NaN               NaN   
91780                  C23                12               200   

      cons_m12_JYDQ_num cons_m12_JYDQ_pay cons_m12_MYYP_num cons_m12_MYYP_pay  
1                   NaN               NaN               NaN               NaN  
2                    55             30000                56              5000  
3                   NaN               NaN                 0                 0  
4                     6               400                12              1000  
5                   NaN               NaN                 0                 0  
...                 ...               ...               ...        

In [236]:
alm_0['cus_num'].to_csv("alm_0_cus_num.csv")
alm_3['cus_num'].to_csv("alm_3_cus_num.csv")

In [229]:
len(result['cus_num'].unique())

88105

In [231]:
len(alm_0['cus_num'].unique())

88105

In [244]:
alm_0['cus_num'] = alm_0['cus_num'].apply(lambda x : str(x))
alm_3['cus_num'] = alm_3['cus_num'].apply(lambda x : str(x))
alm_6['cus_num'] = alm_6['cus_num'].apply(lambda x : str(x))
alm_9['cus_num'] = alm_9['cus_num'].apply(lambda x : str(x))

alm_3_ = alm_3[alm_3['cus_num'].isin(alm_0['cus_num'])]
alm_6_ = alm_6[alm_6['cus_num'].isin(alm_0['cus_num'])]
alm_9_ = alm_9[alm_9['cus_num'].isin(alm_0['cus_num'])]

In [246]:
alm_dict = {}
alm_dict[3] = alm_3_
alm_dict[6] = alm_6_
alm_dict[9] = alm_9_

for t in list(alm_dict.keys()):
    alm_t = alm_dict.get(t)
    zero_counts = {}
    nan_counts = {}
    alm_t_col = alm_t.columns.values
    
    for i in range(len(alm_t_col)):
        vc = alm_t.iloc[:,i].value_counts(dropna=False).to_dict()
        zero_counts[alm_t_col[i]] = vc.get(0) if vc.get(0) else 0 + vc.get("0") if vc.get("0") else 0
        nan_counts[alm_t_col[i]] = alm_t.iloc[:,i].isna().sum()
    
    alm_t_hollow_vc = pd.DataFrame(columns=alm_t_col)
    alm_t_hollow_vc = alm_t_hollow_vc.append(zero_counts,ignore_index=True)
    alm_t_hollow_vc = alm_t_hollow_vc.append(nan_counts,ignore_index=True)
    alm_t_hollow_vc_T = alm_t_hollow_vc.T
    alm_t_hollow_vc_T.columns=['zero_counts','nan_counts']
    alm_t_hollow_vc_T.to_csv("alm_%s_hollow_vc.csv" % t)

# 以下这12个字段的缺失值比例相当高且又存在0值，需要考虑直接fillna(0)的合理性
cons_m3_JYDQ_pay
cons_m3_MYYP_num
cons_m3_MYYP_pay
cons_m3_RYBH_num
cons_m3_RYBH_pay
cons_m12_MYYP_num
cons_m12_MYYP_pay
cons_m12_JYDQ_num
cons_m12_JYDQ_pay
cons_m12_RYBH_num
cons_m12_RYBH_pay

# 目前暂考虑先去除这12个字段，而后考虑fillna(0)对比特征对结果的影响

In [250]:
alm_0.loc[:,'cons_max_m12_paycate'].value_counts(dropna=False).sort_index()

C1      1380
C10     2230
C11     3518
C12       14
C13      518
C14     2558
C15     1706
C16     1555
C17      458
C18     2035
C19      165
C2         3
C20      548
C21      560
C22      659
C23     1729
C24    13486
C25      347
C26     2825
C27      607
C28       12
C29      778
C3       338
C4       164
C5      1970
C7     24180
C8       256
C9      2205
NaN    21301
Name: cons_max_m12_paycate, dtype: int64

## 经过缺失值和0值统计，我们确定了可以作为特征的字段，并开始进行合并处理
* 由于数据是整齐的，可以采用一个4个元素的list来按从近到远，顺序存放

In [251]:
alm_0_col

array(['Unnamed: 0', 'cus_num', 'id', 'cell', 'name', 'flagy',
       'custApiCode', 'user_date', 'swift_number', 'cus_username', 'code',
       'flag_media_c', 'media_tot_m3_visitdays', 'media_tot_m3_catenum',
       'media_max_m3_days', 'media_max_m3_cate', 'media_max_m3_days_prop',
       'media_m3_CJ_visitdays', 'media_m3_WXYS_visitdays',
       'media_m3_MYYE_visitdays', 'media_tot_m12_visitdays',
       'media_tot_m12_catenum', 'media_max_m12_days',
       'media_max_m12_cate', 'media_max_m12_days_prop',
       'media_m12_CJ_visitdays', 'media_m12_WXYS_visitdays',
       'media_m12_MYYE_visitdays', 'flag_consumption_c', 'cons_cont',
       'cons_time_recent', 'cons_tot_m3_visits', 'cons_tot_m3_num',
       'cons_tot_m3_pay', 'cons_tot_m3_p_catenum',
       'cons_tot_m3_v_catenum', 'cons_max_m3_num', 'cons_max_m3_pay',
       'cons_max_m3_paycate', 'cons_m3_RYBH_num', 'cons_m3_RYBH_pay',
       'cons_m3_JYDQ_num', 'cons_m3_JYDQ_pay', 'cons_m3_MYYP_num',
       'cons_m3_MYYP_pay', 

In [ ]:
alm_0

In [301]:
rename_col_3 = {}
old_colunms_3 = alm_3_.columns.values
for i in range(2,len(old_colunms_3)):
    rename_col_3[old_colunms_3[i]] = old_colunms_3[i] + "_3"

In [302]:
alm_3_r = alm_3_.rename(columns=rename_col_3)

In [303]:
rename_col_6 = {}
old_colunms_6 = alm_6_.columns.values
for i in range(2,len(old_colunms_6)):
    rename_col_6[old_colunms_6[i]] = old_colunms_6[i] + "_6"
alm_6_r = alm_6_.rename(columns=rename_col_6)

rename_col_9 = {}
old_colunms_9 = alm_9_.columns.values
for i in range(2,len(old_colunms_9)):
    rename_col_9[old_colunms_9[i]] = old_colunms_9[i] + "_9"
alm_9_r = alm_9_.rename(columns=rename_col_9)

In [304]:
alm_3_r

Unnamed: 0 cus_num                id_3       cell_3 name_3 flagy_3  \
1               1       3  13040219640128##2#  1883106####    冉##       1   
2               2       1  13032419870423##1#  1333333####    宁##       1   
3               3       2  13032419940315##1#  1580117####    闫##       1   
4               4       5  13040219800914##2#  1373004####     张#       1   
5               5       6  13040219820704##4#  1358158####     胡#       1   
...           ...     ...                 ...          ...    ...     ...   
91774       91774   7854b  35062419910225##1#  1885928####    许##       1   
91775       91775   7864b  35062719910818##2#  1330605####     杨#       0   
91776       91776   7868b  35062819810220##6#  1395969####    陈##       0   
91777       91777   7866b  35062819760723##1#  1516053####    庄##       0   
91780       91780   7867b  35062819800916##1#  1396262####    陈##       1   

      custApiCode_3 user_date_3               swift_number_3  cus_username_3  \
1           3000983  2017-07-30  4000056_20190612162224_9236  yejingyou_test   
2           3000983  2017-01-26  4000056_20190612162224_0845  yejingyou_test   
3           3000983  2017-09-04  4000056_20190612162224_2185  yejingyou_test   
4           3000983  2017-05-14  4000056_20190612162224_3083  yejingyou_test   
5           3000983  2017-08-26  4000056_20190612162224_2187  yejingyou_test   
...             ...         ...                          ...             ...   
91774       3100122  2017-04-11  4000056_20190612170225_4877  yejingyou_test   
91775       3100122  2017-04-08  4000056_20190612170225_8347  yejingyou_test   
91776       3100122  2017-04-18  4000056_20190612170225_8349  yejingyou_test   
91777       3100122  2017-04-23  4000056_20190612170225_6153  yejingyou_test   
91780       3100122  2017-04-06  4000056_20190612170225_8348  yejingyou_test   

       ... cons_tot_m12_v_catenum_3 cons_max_m12_num_3 cons_max_m12_pay_3  \
1      ...                      NaN                NaN                NaN   
2      ...                       24                221              30000   
3      ...                        9                 38              10000   
4      ...                       20                250              40000   
5      ...                       20                 34               3000   
...    ...                      ...                ...                ...   
91774  ...                      NaN                NaN                NaN   
91775  ...                       20                148              30000   
91776  ...                       19                 65              10000   
91777  ...                       13                  8               3000   
91780  ...                       10                 17               3000   

      cons_max_m12_paycate_3 cons_m12_RYBH_num_3 cons_m12_RYBH_pay_3  \
1                        NaN                 NaN                 NaN   
2                        C11                 221               20000   
3                        C24                 NaN                 NaN   
4                         C7                   6                  70   
5                         C7                   0                   0   
...                      ...                 ...                 ...   
91774                    NaN                 NaN                 NaN   
91775                     C7                  12                 900   
91776                     C7                  28                 500   
91777                    C26                 NaN                 NaN   
91780                    C23                  12                 200   

      cons_m12_JYDQ_num_3 cons_m12_JYDQ_pay_3 cons_m12_MYYP_num_3  \
1                     NaN                 NaN                 NaN   
2                      55               30000                  56   
3                     NaN                 NaN                   0   
4                       6                 400          

In [305]:
alm_6_r

Unnamed: 0 cus_num                id_6       cell_6 name_6 flagy_6  \
1               1       3  13040219640128##2#  1883106####    冉##       1   
2               2       4  13040219740924##2#  1833006####     李#       1   
3               3       2  13032419940315##1#  1580117####    闫##       1   
4               4       5  13040219800914##2#  1373004####     张#       1   
5               5       1  13032419870423##1#  1333333####    宁##       1   
...           ...     ...                 ...          ...    ...     ...   
91774       91774   7863b  35062719851109##3#  1525979####    余##       1   
91775       91775   7866b  35062819760723##1#  1516053####    庄##       0   
91776       91776   7864b  35062719910818##2#  1330605####     杨#       0   
91777       91777   7868b  35062819810220##6#  1395969####    陈##       0   
91778       91778   7867b  35062819800916##1#  1396262####    陈##       1   

      custApiCode_6 user_date_6               swift_number_6  cus_username_6  \
1           3000983  2017-04-30  4000056_20190612170305_0571  yejingyou_test   
2           3000983  2017-06-08  4000056_20190612170305_8390  yejingyou_test   
3           3000983  2017-06-04  4000056_20190612170305_9295  yejingyou_test   
4           3000983  2017-02-14  4000056_20190612170305_7136  yejingyou_test   
5           3000983  2016-10-26  4000056_20190612170305_9294  yejingyou_test   
...             ...         ...                          ...             ...   
91774       3100122  2017-01-02  4000056_20190612172104_8830  yejingyou_test   
91775       3100122  2017-01-23  4000056_20190612172104_8833  yejingyou_test   
91776       3100122  2017-01-08  4000056_20190612172104_8831  yejingyou_test   
91777       3100122  2017-01-18  4000056_20190612172104_0501  yejingyou_test   
91778       3100122  2017-01-06  4000056_20190612172104_8835  yejingyou_test   

       ... cons_tot_m12_v_catenum_6 cons_max_m12_num_6 cons_max_m12_pay_6  \
1      ...                      NaN                NaN                NaN   
2      ...                        1                  9                  1   
3      ...                        9                 48              10000   
4      ...                       20                358              60000   
5      ...                       24                173              20000   
...    ...                      ...                ...                ...   
91774  ...                        2                 96              10000   
91775  ...                       11                  7               2000   
91776  ...                       19                166              30000   
91777  ...                       19                 54               9000   
91778  ...                       10                 17               3000   

      cons_max_m12_paycate_6 cons_m12_RYBH_num_6 cons_m12_RYBH_pay_6  \
1                        NaN                 NaN                 NaN   
2                        C16                 NaN                 NaN   
3                        C24                 NaN                 NaN   
4                         C7                  24                 500   
5                        C11                 173               20000   
...                      ...                 ...                 ...   
91774                     C7                 NaN                 NaN   
91775                    C26                 NaN                 NaN   
91776                     C7                  17                1000   
91777                    C24                  28                 500   
91778                    C23                   4                  70   

      cons_m12_JYDQ_num_6 cons_m12_JYDQ_pay_6 cons_m12_MYYP_num_6  \
1                     NaN                 NaN                 NaN   
2                     NaN                 NaN                 NaN   
3                     NaN                 NaN                   0   
4                      14                 800          

In [306]:
alm_9_r

Unnamed: 0 cus_num                id_9       cell_9 name_9 flagy_9  \
1               1       3  13040219640128##2#  1883106####    冉##       1   
2               2       2  13032419940315##1#  1580117####    闫##       1   
3               3       4  13040219740924##2#  1833006####     李#       1   
4               4       1  13032419870423##1#  1333333####    宁##       1   
5               5       5  13040219800914##2#  1373004####     张#       1   
...           ...     ...                 ...          ...    ...     ...   
91774       91774   7855b  35062419910424##1#  1340066####    沈##       0   
91775       91775   7866b  35062819760723##1#  1516053####    庄##       0   
91776       91776   7864b  35062719910818##2#  1330605####     杨#       0   
91777       91777   7867b  35062819800916##1#  1396262####    陈##       1   
91778       91778   7868b  35062819810220##6#  1395969####    陈##       0   

      custApiCode_9 user_date_9               swift_number_9  cus_username_9  \
1           3000983  2017-01-30  4000056_20190612172123_3882  yejingyou_test   
2           3000983  2017-03-04  4000056_20190612172123_1731  yejingyou_test   
3           3000983  2017-03-08  4000056_20190612172123_3883  yejingyou_test   
4           3000983  2016-07-26  4000056_20190612172123_5173  yejingyou_test   
5           3000983  2016-11-14  4000056_20190612172123_1730  yejingyou_test   
...             ...         ...                          ...             ...   
91774       3100122  2016-10-12  4000056_20190612173904_8900  yejingyou_test   
91775       3100122  2016-10-23  4000056_20190612173904_1209  yejingyou_test   
91776       3100122  2016-10-08  4000056_20190612173904_1208  yejingyou_test   
91777       3100122  2016-10-06  4000056_20190612173904_8904  yejingyou_test   
91778       3100122  2016-10-18  4000056_20190612173904_7235  yejingyou_test   

       ... cons_tot_m12_v_catenum_9 cons_max_m12_num_9 cons_max_m12_pay_9  \
1      ...                      NaN                NaN                NaN   
2      ...                        9                 61              10000   
3      ...                        1                  9                  1   
4      ...                       24                107              10000   
5      ...                       20                348              50000   
...    ...                      ...                ...                ...   
91774  ...                       24                 68               4000   
91775  ...                       11                  5               2000   
91776  ...                       19                139              20000   
91777  ...                        9                  3                700   
91778  ...                       19                 50               9000   

      cons_max_m12_paycate_9 cons_m12_RYBH_num_9 cons_m12_RYBH_pay_9  \
1                        NaN                 NaN                 NaN   
2                        C24                 NaN                 NaN   
3                        C16                 NaN                 NaN   
4                        C17                 107               10000   
5                         C7                  24                 500   
...                      ...                 ...                 ...   
91774                     C7                  68                2000   
91775                    C26                 NaN                 NaN   
91776                     C7                  16                1000   
91777                    C23                   1                  20   
91778                    C24                  28                 500   

      cons_m12_JYDQ_num_9 cons_m12_JYDQ_pay_9 cons_m12_MYYP_num_9  \
1                     NaN                 NaN                 NaN   
2                     NaN                 NaN                   0   
3                     NaN                 NaN                 NaN   
4                      39                6000          

In [307]:
rename_col_0 = {}
old_colunms_0 = alm_0.columns.values
for i in range(2,len(old_colunms_0)):
    rename_col_0[old_colunms_0[i]] = old_colunms_0[i] + "_0"
alm_0_r = alm_0.rename(columns=rename_col_0)

In [309]:
alm_0_r

Unnamed: 0 cus_num                id_0       cell_0 name_0  flagy_0  \
0               0       1  13032419870423##1#  1333333####    宁##        1   
1               1      10  13040219940711##2#  1508100####    蔺##        1   
2               2     100  13043119850415##3#  1383208####     桑#        1   
3               3    1000  13092519900220##2#  1346376####    刘##        1   
4               4   10000  14270319660329##4#  1363359####    吴##        0   
...           ...     ...                 ...          ...    ...      ...   
88100       88100    7869  32083119931014##3#  1865240####    闵##        0   
88101       88101   78690  61032619910208##6#  1839276####    刘##        0   
88102       88102   78691  61010419990302##1#  1336399####    张##        0   
88103       88103   78692  45092319900504##2#  1859394####    刘##        1   
88104       88104   78693  42900520000719##1#  1354596####    肖##        0   

       custApiCode_0 user_date_0               swift_number_0  cus_username_0  \
0            3000983  2017-04-26  4000056_20190606101948_6234  yejingyou_test   
1            3000983  2017-10-26  4000056_20190606101948_3022  yejingyou_test   
2            3000983  2017-12-09  4000056_20190606101952_4792  yejingyou_test   
3            3000983  2017-12-05  4000056_20190606102032_3642  yejingyou_test   
4            3000983  2017-05-10  4000056_20190606102626_0291  yejingyou_test   
...              ...         ...                          ...             ...   
88100        3000983  2018-01-24  4000056_20190606102504_9565  yejingyou_test   
88101        3000013  2018-08-09  4000056_20190606110921_3412  yejingyou_test   
88102        3000013  2018-08-09  4000056_20190606110921_3413  yejingyou_test   
88103        3000013  2018-08-09  4000056_20190606110921_3414  yejingyou_test   
88104        3000013  2018-08-09  4000056_20190606110921_0241  yejingyou_test   

       ...  cons_tot_m12_v_catenum_0  cons_max_m12_num_0  cons_max_m12_pay_0  \
0      ...                      24.0               183.0             30000.0   
1      ...                      19.0                35.0              3000.0   
2      ...                      17.0                10.0               600.0   
3      ...                      17.0               216.0             40000.0   
4      ...                       9.0                46.0              4000.0   
...    ...                       ...                 ...                 ...   
88100  ...                      17.0                78.0             20000.0   
88101  ...                      11.0                 4.0               600.0   
88102  ...                       NaN                 NaN                 NaN   
88103  ...                       NaN                 NaN                 NaN   
88104  ...                      10.0                 6.0              2000.0   

       cons_max_m12_paycate_0  cons_m12_RYBH_num_0 cons_m12_RYBH_pay_0  \
0                         C11                183.0             10000.0   
1                          C7                  2.0                20.0   
2                         C11                  0.0                 0.0   
3                          C7                 11.0              2000.0   
4                         C14                  NaN                 NaN   
...                       ...                  ...                 ...   
88100                      C7                  1.0                30.0   
88101                     C29                  0.0                 0.0   
88102                     NaN                  NaN                 NaN   
88103                     NaN                  NaN                 NaN   
88104                     C23                  0.0                 0.0   

       cons_m12_JYDQ_num_0  cons_m12_JYDQ_pay_0  cons_m12_MYYP_num_0  \
0                     52.0              30000.0                 38.0   
1                      0.0                  0.0                 10.0   
2                      3.0     

In [310]:
# 将四个重命名后的alm_i_r进行数据类型转换，以便合并

alm_all_r = pd.merge(alm_0_r, alm_3_r, how='left', on=['cus_num'])
alm_all_r = pd.merge(alm_all_r, alm_6_r, how='left', on=['cus_num'])
alm_all_r = pd.merge(alm_all_r, alm_9_r, how='left', on=['cus_num'])

In [311]:
alm_all_r

Unnamed: 0_x cus_num                id_0       cell_0 name_0  flagy_0  \
0                 0       1  13032419870423##1#  1333333####    宁##        1   
1                 1      10  13040219940711##2#  1508100####    蔺##        1   
2                 2     100  13043119850415##3#  1383208####     桑#        1   
3                 3    1000  13092519900220##2#  1346376####    刘##        1   
4                 4   10000  14270319660329##4#  1363359####    吴##        0   
...             ...     ...                 ...          ...    ...      ...   
88100         88100    7869  32083119931014##3#  1865240####    闵##        0   
88101         88101   78690  61032619910208##6#  1839276####    刘##        0   
88102         88102   78691  61010419990302##1#  1336399####    张##        0   
88103         88103   78692  45092319900504##2#  1859394####    刘##        1   
88104         88104   78693  42900520000719##1#  1354596####    肖##        0   

       custApiCode_0 user_date_0               swift_number_0  cus_username_0  \
0            3000983  2017-04-26  4000056_20190606101948_6234  yejingyou_test   
1            3000983  2017-10-26  4000056_20190606101948_3022  yejingyou_test   
2            3000983  2017-12-09  4000056_20190606101952_4792  yejingyou_test   
3            3000983  2017-12-05  4000056_20190606102032_3642  yejingyou_test   
4            3000983  2017-05-10  4000056_20190606102626_0291  yejingyou_test   
...              ...         ...                          ...             ...   
88100        3000983  2018-01-24  4000056_20190606102504_9565  yejingyou_test   
88101        3000013  2018-08-09  4000056_20190606110921_3412  yejingyou_test   
88102        3000013  2018-08-09  4000056_20190606110921_3413  yejingyou_test   
88103        3000013  2018-08-09  4000056_20190606110921_3414  yejingyou_test   
88104        3000013  2018-08-09  4000056_20190606110921_0241  yejingyou_test   

       ...  cons_tot_m12_v_catenum_9  cons_max_m12_num_9  cons_max_m12_pay_9  \
0      ...                        24                 107               10000   
1      ...                        22                 138               20000   
2      ...                        15                  14                 800   
3      ...                        15                 285               50000   
4      ...                         7                   2                  80   
...    ...                       ...                 ...                 ...   
88100  ...                        21                  96               20000   
88101  ...                        10                   4                 600   
88102  ...                       NaN                 NaN                 NaN   
88103  ...                       NaN                 NaN                 NaN   
88104  ...                        10                   6                2000   

       cons_max_m12_paycate_9  cons_m12_RYBH_num_9 cons_m12_RYBH_pay_9  \
0                         C17                  107               10000   
1                          C7                   16                 600   
2                          C5                    0                   0   
3                          C7                    7                2000   
4                         C26                  NaN                 NaN   
...                       ...                  ...                 ...   
88100                      C7                    3                  70   
88101                     C29                    0                   0   
88102                     NaN                  NaN                 NaN   
88103                     NaN                  NaN                 NaN   
88104                     C23                    0                   0   

       cons_m12_JYDQ_num_9  cons_m12_JYDQ_pay_9  cons_m12_MYYP_num_9  \
0                       39                 6000                   42   
1                        4                   90                   12   
2      

In [336]:
alm_all_r.sort_index(axis=1,inplace=True)

In [314]:
# 开始进行合并处理
origin_col = alm_0_col

In [317]:
empty_df = pd.DataFrame(columns=origin_col)

In [341]:
# def call_in(a, b, c, d):
#     return [a,b,c,d]
 
# empty_df['cell'] = alm_all_r.apply(lambda r: call_in(r['cell_0'], r['cell_3'],r['cell_6'],r['cell_9']),axis=1)
# 去除掉cun_num
col_ = alm_all_r.columns.values
for col_name in col_:
    # 去除尾部的"_i"
    if col_name not in ['cus_num','Unnamed: 0_x', 'Unnamed: 0_x', 'Unnamed: 0_y', 'Unnamed: 0_y']: 
        empty_df[col_name] = alm_all_r.apply(lambda r: [r[col_name[:-2]+"_%d"%i] for i in [0,3,6,9]],axis=1)

In [342]:
empty_df

Unnamed: 0 cus_num   id                                               cell  \
0        NaN     NaN  NaN  [1333333####, 1333333####, 1333333####, 133333...   
1        NaN     NaN  NaN  [1508100####, 1508100####, 1508100####, 150810...   
2        NaN     NaN  NaN  [1383208####, 1383208####, 1383208####, 138320...   
3        NaN     NaN  NaN  [1346376####, 1346376####, 1346376####, 134637...   

  name flagy custApiCode user_date swift_number cus_username  ...  \
0  NaN   NaN         NaN       NaN          NaN          NaN  ...   
1  NaN   NaN         NaN       NaN          NaN          NaN  ...   
2  NaN   NaN         NaN       NaN          NaN          NaN  ...   
3  NaN   NaN         NaN       NaN          NaN          NaN  ...   

                 name_6                name_9  \
0  [宁##, 宁##, 宁##, 宁##]  [宁##, 宁##, 宁##, 宁##]   
1  [蔺##, 蔺##, 蔺##, 蔺##]  [蔺##, 蔺##, 蔺##, 蔺##]   
2      [桑#, 桑#, 桑#, 桑#]      [桑#, 桑#, 桑#, 桑#]   
3  [刘##, 刘##, 刘##, 刘##]  [刘##, 刘##, 刘##, 刘##]   

                                      swift_number_0  \
0  [4000056_20190606101948_6234, 4000056_20190612...   
1  [4000056_20190606101948_3022, 4000056_20190612...   
2  [4000056_20190606101952_4792, 4000056_20190612...   
3  [4000056_20190606102032_3642, 4000056_20190612...   

                                      swift_number_3  \
0  [4000056_20190606101948_6234, 4000056_20190612...   
1  [4000056_20190606101948_3022, 4000056_20190612...   
2  [4000056_20190606101952_4792, 4000056_20190612...   
3  [4000056_20190606102032_3642, 4000056_20190612...   

                                      swift_number_6  \
0  [4000056_20190606101948_6234, 4000056_20190612...   
1  [4000056_20190606101948_3022, 4000056_20190612...   
2  [4000056_20190606101952_4792, 4000056_20190612...   
3  [4000056_20190606102032_3642, 4000056_20190612...   

                                      swift_number_9  \
0  [4000056_20190606101948_6234, 4000056_20190612...   
1  [4000056_20190606101948_3022, 4000056_20190612...   
2  [4000056_20190606101952_4792, 4000056_20190612...   
3  [4000056_20190606102032_3642, 4000056_20190612...   

                                        user_date_0  \
0  [2017-04-26, 2017-01-26, 2016-10-26, 2016-07-26]   
1  [2017-10-26, 2017-07-26, 2017-04-26, 2017-01-26]   
2  [2017-12-09, 2017-09-09, 2017-06-09, 2017-03-09]   
3  [2017-12-05, 2017-09-05, 2017-06-05, 2017-03-05]   

                                        user_date_3  \
0  [2017-04-26, 2017-01-26, 2016-10-26, 2016-07-26]   
1  [2017-10-26, 2017-07-26, 2017-04-26, 2017-01-26]   
2  [2017-12-09, 2017-09-09, 2017-06-09, 2017-03-09]   
3  [2017-12-05, 2017-09-05, 2017-06-05, 2017-03-05]   

                                        user_date_6  \
0  [2017-04-26, 2017-01-26, 2016-10-26, 2016-07-26]   
1  [2017-10-26, 2017-07-26, 2017-04-26, 2017-01-26]   
2  [2017-12-09, 2017-09-09, 2017-06-09, 2017-03-09]   
3  [2017-12-05, 2017-09-05, 2017-06-05, 2017-03-05]   

                                        user_date_9  
0  [2017-04-26, 2017-01-26, 2016-10-26, 2016-07-26]  
1  [2017-10-26, 2017-07-26, 2017-04-26, 2017-01-26]  
2  [2017-12-09, 2017-09-09, 2017-06-09, 2017-03-09]  
3  [2017-12-05, 2017-09-05, 2017-06-05, 2017-03-05]  

[4 rows x 287 columns]

In [343]:
# def call_in(a, b, c, d):
#     return [a,b,c,d]
 
# empty_df['cell'] = alm_all_r.apply(lambda r: call_in(r['cell_0'], r['cell_3'],r['cell_6'],r['cell_9']),axis=1)
# 去除掉cun_num
empty_df_ = pd.DataFrame(columns=origin_col)
col_ = alm_all_r.columns.values
for col_name in col_:
    # 去除尾部的"_i"
    if col_name not in ['cus_num','Unnamed: 0_x', 'Unnamed: 0_x', 'Unnamed: 0_y', 'Unnamed: 0_y']: 
        empty_df_[col_name[:-2]] = alm_all_r.apply(lambda r: [r[col_name[:-2]+"_%d"%i] for i in [0,3,6,9]],axis=1)

In [344]:
empty_df_

Unnamed: 0 cus_num                                                 id  \
0            NaN     NaN  [13032419870423##1#, 13032419870423##1#, 13032...   
1            NaN     NaN  [13040219940711##2#, 13040219940711##2#, 13040...   
2            NaN     NaN  [13043119850415##3#, 13043119850415##3#, 13043...   
3            NaN     NaN  [13092519900220##2#, 13092519900220##2#, 13092...   
4            NaN     NaN  [14270319660329##4#, 14270319660329##4#, 14270...   
...          ...     ...                                                ...   
88100        NaN     NaN  [32083119931014##3#, 32083119931014##3#, 32083...   
88101        NaN     NaN  [61032619910208##6#, 61032619910208##6#, 61032...   
88102        NaN     NaN  [61010419990302##1#, 61010419990302##1#, 61010...   
88103        NaN     NaN  [45092319900504##2#, 45092319900504##2#, 45092...   
88104        NaN     NaN  [42900520000719##1#, 42900520000719##1#, 42900...   

                                                    cell  \
0      [1333333####, 1333333####, 1333333####, 133333...   
1      [1508100####, 1508100####, 1508100####, 150810...   
2      [1383208####, 1383208####, 1383208####, 138320...   
3      [1346376####, 1346376####, 1346376####, 134637...   
4      [1363359####, 1363359####, 1363359####, 136335...   
...                                                  ...   
88100  [1865240####, 1865240####, 1865240####, 186524...   
88101  [1839276####, 1839276####, 1839276####, 183927...   
88102  [1336399####, 1336399####, 1336399####, 133639...   
88103  [1859394####, 1859394####, 1859394####, 185939...   
88104  [1354596####, 1354596####, 1354596####, 135459...   

                       name         flagy  \
0      [宁##, 宁##, 宁##, 宁##]  [1, 1, 1, 1]   
1      [蔺##, 蔺##, 蔺##, 蔺##]  [1, 1, 1, 1]   
2          [桑#, 桑#, 桑#, 桑#]  [1, 1, 1, 1]   
3      [刘##, 刘##, 刘##, 刘##]  [1, 1, 1, 1]   
4      [吴##, 吴##, 吴##, 吴##]  [0, 0, 0, 0]   
...                     ...           ...   
88100  [闵##, 闵##, 闵##, 闵##]  [0, 0, 0, 0]   
88101  [刘##, 刘##, 刘##, 刘##]  [0, 0, 0, 0]   
88102  [张##, 张##, 张##, 张##]  [0, 0, 0, 0]   
88103  [刘##, 刘##, 刘##, 刘##]  [1, 1, 1, 1]   
88104  [肖##, 肖##, 肖##, 肖##]  [0, 0, 0, 0]   

                                custApiCode  \
0      [3000983, 3000983, 3000983, 3000983]   
1      [3000983, 3000983, 3000983, 3000983]   
2      [3000983, 3000983, 3000983, 3000983]   
3      [3000983, 3000983, 3000983, 3000983]   
4      [3000983, 3000983, 3000983, 3000983]   
...                                     ...   
88100  [3000983, 3000983, 3000983, 3000983]   
88101  [3000013, 3000013, 3000013, 3000013]   
88102  [3000013, 3000013, 3000013, 3000013]   
88103  [3000013, 3000013, 3000013, 3000013]   
88104  [3000013, 3000013, 3000013, 3000013]   

                                              user_date  \
0      [2017-04-26, 2017-01-26, 2016-10-26, 2016-07-26]   
1      [2017-10-26, 2017-07-26, 2017-04-26, 2017-01-26]   
2      [2017-12-09, 2017-09-09, 2017-06-09, 2017-03-09]   
3      [2017-12-05, 2017-09-05, 2017-06-05, 2017-03-05]   
4      [2017-05-10, 2017-02-10, 2016-11-10, 2016-08-10]   
...                                                 ...   
88100  [2018-01-24, 2017-10-24, 2017-07-24, 2017-04-24]   
88101  [2018-08-09, 2018-05-09, 2018-02-09, 2017-11-09]   
88102  [2018-08-09, 2018-05-09, 2018-02-09, 2017-11-09]   
88103  [2018-08-09, 2018-05-09, 2018-02-09, 2017-11-09]   
88104  [2018-08-09, 2018-05-09, 2018-02-09, 2017-11-09]   

                                            swift_number  \
0      [4000056_20190606101948_6234, 4000056_20190612...   
1      [4000056_20190606101948_3022, 4000056_20190612...   
2      [4000056_20190606101952_4792, 4000056_20190612...   
3      [4000056_20190606102032_3642, 4000056_20190612...   
4      [4000056_20190606102626_0291, 4000056_20190612...   
...                                                  ...   
88100  [4000056_20190606102504_9565, 4000056_20190612...   
88101  [4000056_20190606110921_3412, 4000

In [358]:
alm_tmp = alm_all_r.loc[:,['cus_num','id_0']]

In [359]:
alm_tmp['id_0'] = alm_tmp.apply(lambda r: [r['id_0'] for i in range(4)],axis=1)

In [364]:
alm_tmp.rename(columns={'id_0':'id'},inplace=True)

In [365]:
alm_tmp

cus_num                                                 id
0           1  [13032419870423##1#, 13032419870423##1#, 13032...
1          10  [13040219940711##2#, 13040219940711##2#, 13040...
2         100  [13043119850415##3#, 13043119850415##3#, 13043...
3        1000  [13092519900220##2#, 13092519900220##2#, 13092...
4       10000  [14270319660329##4#, 14270319660329##4#, 14270...
...       ...                                                ...
88100    7869  [32083119931014##3#, 32083119931014##3#, 32083...
88101   78690  [61032619910208##6#, 61032619910208##6#, 61032...
88102   78691  [61010419990302##1#, 61010419990302##1#, 61010...
88103   78692  [45092319900504##2#, 45092319900504##2#, 45092...
88104   78693  [42900520000719##1#, 42900520000719##1#, 42900...

[88105 rows x 2 columns]

In [367]:
# res = pd.merge(empty_df_, alm_tmp, how='left', on=['id'])
alm_tmp['id'] = alm_tmp.apply(lambda r: ",".join(r['id']), axis=1)

In [370]:
empty_df_['id'] = empty_df_.apply(lambda r: ",".join(r['id']), axis=1)

In [373]:
alm_tmp

cus_num                                                 id
0           1  13032419870423##1#,13032419870423##1#,13032419...
1          10  13040219940711##2#,13040219940711##2#,13040219...
2         100  13043119850415##3#,13043119850415##3#,13043119...
3        1000  13092519900220##2#,13092519900220##2#,13092519...
4       10000  14270319660329##4#,14270319660329##4#,14270319...
...       ...                                                ...
88100    7869  32083119931014##3#,32083119931014##3#,32083119...
88101   78690  61032619910208##6#,61032619910208##6#,61032619...
88102   78691  61010419990302##1#,61010419990302##1#,61010419...
88103   78692  45092319900504##2#,45092319900504##2#,45092319...
88104   78693  42900520000719##1#,42900520000719##1#,42900520...

[88105 rows x 2 columns]

In [385]:
empty_df_['cus_num'] = alm_all_r['cus_num']

In [387]:
empty_df_.drop(columns=['Unnamed: 0'])

cus_num                                                 id  \
0           1  13032419870423##1#,13032419870423##1#,13032419...   
1          10  13040219940711##2#,13040219940711##2#,13040219...   
2         100  13043119850415##3#,13043119850415##3#,13043119...   
3        1000  13092519900220##2#,13092519900220##2#,13092519...   
4       10000  14270319660329##4#,14270319660329##4#,14270319...   
...       ...                                                ...   
88100    7869  32083119931014##3#,32083119931014##3#,32083119...   
88101   78690  61032619910208##6#,61032619910208##6#,61032619...   
88102   78691  61010419990302##1#,61010419990302##1#,61010419...   
88103   78692  45092319900504##2#,45092319900504##2#,45092319...   
88104   78693  42900520000719##1#,42900520000719##1#,42900520...   

                                                    cell  \
0      [1333333####, 1333333####, 1333333####, 133333...   
1      [1508100####, 1508100####, 1508100####, 150810...   
2      [1383208####, 1383208####, 1383208####, 138320...   
3      [1346376####, 1346376####, 1346376####, 134637...   
4      [1363359####, 1363359####, 1363359####, 136335...   
...                                                  ...   
88100  [1865240####, 1865240####, 1865240####, 186524...   
88101  [1839276####, 1839276####, 1839276####, 183927...   
88102  [1336399####, 1336399####, 1336399####, 133639...   
88103  [1859394####, 1859394####, 1859394####, 185939...   
88104  [1354596####, 1354596####, 1354596####, 135459...   

                       name         flagy  \
0      [宁##, 宁##, 宁##, 宁##]  [1, 1, 1, 1]   
1      [蔺##, 蔺##, 蔺##, 蔺##]  [1, 1, 1, 1]   
2          [桑#, 桑#, 桑#, 桑#]  [1, 1, 1, 1]   
3      [刘##, 刘##, 刘##, 刘##]  [1, 1, 1, 1]   
4      [吴##, 吴##, 吴##, 吴##]  [0, 0, 0, 0]   
...                     ...           ...   
88100  [闵##, 闵##, 闵##, 闵##]  [0, 0, 0, 0]   
88101  [刘##, 刘##, 刘##, 刘##]  [0, 0, 0, 0]   
88102  [张##, 张##, 张##, 张##]  [0, 0, 0, 0]   
88103  [刘##, 刘##, 刘##, 刘##]  [1, 1, 1, 1]   
88104  [肖##, 肖##, 肖##, 肖##]  [0, 0, 0, 0]   

                                custApiCode  \
0      [3000983, 3000983, 3000983, 3000983]   
1      [3000983, 3000983, 3000983, 3000983]   
2      [3000983, 3000983, 3000983, 3000983]   
3      [3000983, 3000983, 3000983, 3000983]   
4      [3000983, 3000983, 3000983, 3000983]   
...                                     ...   
88100  [3000983, 3000983, 3000983, 3000983]   
88101  [3000013, 3000013, 3000013, 3000013]   
88102  [3000013, 3000013, 3000013, 3000013]   
88103  [3000013, 3000013, 3000013, 3000013]   
88104  [3000013, 3000013, 3000013, 3000013]   

                                              user_date  \
0      [2017-04-26, 2017-01-26, 2016-10-26, 2016-07-26]   
1      [2017-10-26, 2017-07-26, 2017-04-26, 2017-01-26]   
2      [2017-12-09, 2017-09-09, 2017-06-09, 2017-03-09]   
3      [2017-12-05, 2017-09-05, 2017-06-05, 2017-03-05]   
4      [2017-05-10, 2017-02-10, 2016-11-10, 2016-08-10]   
...                                                 ...   
88100  [2018-01-24, 2017-10-24, 2017-07-24, 2017-04-24]   
88101  [2018-08-09, 2018-05-09, 2018-02-09, 2017-11-09]   
88102  [2018-08-09, 2018-05-09, 2018-02-09, 2017-11-09]   
88103  [2018-08-09, 2018-05-09, 2018-02-09, 2017-11-09]   
88104  [2018-08-09, 2018-05-09, 2018-02-09, 2017-11-09]   

                                            swift_number  \
0      [4000056_20190606101948_6234, 4000056_20190612...   
1      [4000056_20190606101948_3022, 4000056_20190612...   
2      [4000056_20190606101952_4792, 4000056_20190612...   
3      [4000056_20190606102032_3642, 4000056_20190612...   
4      [4000056_20190606102626_0291, 4000056_20190612...   
...                                                  ...   
88100  [4000056_20190606102504_9565, 4000056_20190612...   
88101  [4000056_20190606110921_3412, 4000056_20190612...   
88102  [4000056_20190606110921_3413, 4000056_20190612...   
88103  [4000056_20190606110921_3414, 4000056_20190612

In [ ]:

​